multinet/api.py

```diff
@@ -24,7 +27,31 @@
 @swag_from("swagger/workspaces.yaml")
 def get_workspaces() -> Any:
     """Retrieve list of workspaces."""
-    return util.stream(db.get_workspaces())
+    user = session_user()
+    current_app.logger.info(user)
+
+    def can_read(user: Optional[UserInfo], ws: Workspace) -> bool:
+        perms = ws["permissions"]
+
+        # No one logged in means only return the public workspaces.
+        if user is None:
+            return perms["public"]
+
+        # Otherwise, check to see if the workspace is public, or the user is at
+        # least a Reader of the workspace.
+        sub = user.sub
+        return (
+            perms["public"]
+            or perms["owner"] == sub
+            or sub in perms["readers"]
+            or sub in perms["writers"]
+            or sub in perms["maintainers"]
+        )
```

Comment:
I think this would be better placed outside of `get_workspaces` (and probably renamed to something like `workspace_readable`), as nothing in this function relies on anything besides the input parameters. 

Alternatively, we could make a general `workspace_permissions` function (name not withstanding), which would take a workspace and a user, but would instead return all of the permissions the user has on that workspace, instead of just read, just write, etc. This would have a new signature that describes a user's permissions on a workspace, something like:

```
{
    "owner": bool,
    "reader": bool,
    "writer": bool,
    "maintainer": bool
}
```

Then below, instead of `can_read(user, w)`, you'd replace it with `workspace_permissions(user, w)["reader"]`.

I'd also think that if we were to create this function, we'd put it in either `db.py` or `util.py`.

Reply:
In our permissions model, each role includes implicitly the roles that are "weaker"; that is, an owner of a workspace is also, by definition, a maintainer, writer, and reader. Since that's the case, and our needs are for functions that can filter by these roles, I'm going to create a collection of `-> bool` functions for each role, and create associated decorators to guard the use of API endpoints (in this case, we wouldn't use the decorator, since we're filtering all workspaces by that hypothetical `is_reader()` function).

I will push a commit shortly to do a bit of refactoring here.

Reply:
See 282cbf4101ae3b4e0a221453ce3f82a6bc16f997.